lib/seahorn/CexExeGenerator.cc

```diff
@@ -37,122 +37,133 @@ bool extractArrayContents(Expr e, kv &out, Expr &defaultValue) {
-  } else {
-    LOG("cex",
-        errs() << "unhandled Value of memhavoc ptr: " << *hPtrAlloc << "\n");
-    return;
+class BmcTraceVisitor : public InstVisitor<BmcTraceVisitor<Trace>> {
+  CexExeGenerator<Trace> &m_cex;
+  unsigned m_loc;
+
+public:
+  BmcTraceVisitor(CexExeGenerator<Trace> &cex) : m_cex(cex) {}
+
+  void setLoc(unsigned loc) { m_loc = loc; }
+
+  void visitMemhavoc(CallSite CS) {
+    Instruction &I = *CS.getInstruction();
+    const Value *CV = CS.getCalledValue();
+    const Function *CF = dyn_cast<Function>(CV->stripPointerCasts());
+    // previous instruction should be
+    // shadow.mem.load(i32 x, i32 %sm_n, i8* null)
+    if (I.getPrevNonDebugInstruction() == nullptr)
+      return;
+    const Instruction *prevI = I.getPrevNonDebugInstruction();
+    if (const CallInst *prevCI = dyn_cast<CallInst>(prevI)) {
+      ImmutableCallSite prevCS(prevCI);
+      const Value *prevCV = prevCS.getCalledValue();
+      const Function *prevF = dyn_cast<Function>(prevCV->stripPointerCasts());
+      if (!(prevF && prevF->getName().equals("shadow.mem.load"))) {
+        LOG("cex", errs() << "Skipping harness for memhavoc"
+                          << " because shadow.mem.load cannot be found\n");
+        return;
+      }
+      // get memhavoc content from corresponding shadow mem region
+      auto *shadowMemI = dyn_cast<Instruction>(prevCS.getArgOperand(1));
+      if (!shadowMemI)
+        return;
+      Expr shadowMem = m_cex.trace().eval(m_loc, *shadowMemI, true);
+      // get memhavoc size from second operand of memhavoc
+      const Value *sizeArg = CS.getArgOperand(1);
+      Expr size;
+      if (auto *sizeI = dyn_cast<Instruction>(sizeArg)) {
+        size = m_cex.trace().eval(m_loc, *sizeI, true);
+      } else if (auto *sizeConst = dyn_cast<ConstantInt>(sizeArg)) {
+        expr::mpz_class sz = toMpz(sizeConst);
+        size = expr::mkTerm<expr::mpz_class>(sz, m_cex.trace().engine().efac());
+      } else {
+        LOG("cex",
+            errs() << "unhandled Value of memhavoc size: " << *sizeArg << "\n");
+        return;
+      }
+      // get info of ptr to havoc
+      const Value *hPtrAlloc = CS.getArgOperand(0)->stripPointerCasts();
+      Expr hPtrStart;
+      if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtrAlloc)) {
+        hPtrStart = m_cex.trace().eval(m_loc, *hPtrAllocInst, true);
+      } else {
+        LOG("cex", errs() << "unhandled Value of memhavoc ptr: " << *hPtrAlloc
+                          << "\n");
+        return;
+      }
+      Expr shadowMemRaw = m_cex.trace().engine().getPtrAddressable(shadowMem);
+      Expr hPtrStartRaw = m_cex.trace().engine().getPtrAddressable(hPtrStart);
+      if (!shadowMemRaw || !hPtrStartRaw || !size) {
+        LOG("cex",
+            errs() << "Skipping harness for memhavoc due to lacking info \n");
+        return;
+      }
+      LOG("cex", errs() << "Producing harness for " << CF->getName() << "\n";);
+      m_cex.addValueToFunc(CF, shadowMemRaw);
```

Comment:
I do not understand `shadowMemRaw`. The argument to `shadow.mem.load` is a version of a memory region. it is not a pointer... I must be missing something.

Reply:
in sea-dsa we injected `shadow.mem.load` in front of every `memhavoc` to associate every havoc pointer with corresponding shadow mem region; the idea is to evaluate the memory region for memory content.

Reply:
I think I finally understand what I think is wrong. See comments above about `shadowMem`